### DISCUSSION TITLE
How to convert camera image format (such as. Nef) to other formats without changing image color？

### DISCUSSION THREAD

**Discussion by xiaohenian** [ImageMagick/ImageMagick#3161] [Help]
Through the command line to set the camera white balance and change the image color space can achieve my needs：
`magick -define dng:use-camera-wb=true -define dng:output-color=6 _DSC4324.NEF 1.jpg`
But I want to do it in other ways. For example, by modifying“ delegates.xml ”File, or through C++.

**Comment by xiaohenian** [comment]
Now when I convert the Nef image directly to another format image, the new image color turns red.

**Comment by snibgo** [comment]
> -define dng:output-color=6

This converts to the ACES colorspace. This is fine, if your subsequent workflow is ACES, which is used for high-precision accurate colours. But you make a JPG which is low-precision and lossy.

If you remove "-define dng:output-color=6", the result will be sRGB, which might suit your needs better.

> ...  the new image color turns red.

How do you know? Perhaps your display program assumes the image is sRGB.

> ... or through C++.

ImageMagick has a number of APIs, including C++. IM has sample CPP programs that can get you started.

**Reply by xiaohenian** [reply]
If it is not converted to ACES, the Nef image will be converted to other image formats, and the image will change color. Sorry, the sample image is too big  to upload.

**Reply by xiaohenian** [reply]
I view it through the picture viewer of windows 10.

**Comment by snibgo** [comment]
> If it is not converted to ACES, the Nef image will be converted to other image formats, and the image will change color.

ACES is not an image format. It is a colorspace. Why do you want the image to be encoded as ACES colorspace?

**Comment by xiaohenian** [comment]
Sorry, the sample image is too big to upload.So I can only upload screenshots.
This is what Windows 10 looks like when it opens this NEF image：
![1](https://user-images.githubusercontent.com/37136059/105936606-f65f5f00-608e-11eb-9033-d8af6ca64abc.png)
But if I didn't encode it as ACES color space, the resulting image looks like this：
![2](https://user-images.githubusercontent.com/37136059/105936766-40e0db80-608f-11eb-972c-b2f1eb51f406.png)

**Comment by xiaohenian** [comment]
[FIS_0233.zip](https://github.com/ImageMagick/ImageMagick/files/5879575/FIS_0233.zip)
This is the original image file.

**Comment by xiaohenian** [comment]
[FIS_0233_png.zip](https://github.com/ImageMagick/ImageMagick/files/5879602/FIS_0233_png.zip)
This is the file converted to PNG format.It is obvious that the color has changed after the conversion. How to make the converted image as close as possible to the original image?

**Comment by snibgo** [comment]
Perhaps the "picture viewer of windows 10" shows the embedded JPG rather than processing the raw image. Or perhaps it uses a different algorithm for white balance. Or perhaps your version of IM is too old, before "-define dng:use-camera-wb=true" was implemented.

Please show the text output from "magick -version".

FIS_0233.zip contains an entirely different image.

**Reply by xiaohenian** [reply]
FIS_0233.zip contains another NEF image. I found that as long as the image is in NEF format, the image color will change significantly after being converted to other formats. As I posted above, fis_0233.nef and fis_0233.png.

**Comment by xiaohenian** [comment]
`Version: ImageMagick 7.0.10-38 Q8 x86 2020-12-22 http://www.imagemagick.org`
`Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC`
`License: http://www.imagemagick.org/script/license.php`
`Visual C++: 191627032`
`Features: Cipher DPC HDRI OpenCL OpenMP(2.0)`
`Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr
 lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib`

**Comment by snibgo** [comment]
What is the problem with FIS_0233?

The embedded JPG has similar colours to a version created by IM from the raw image. The JPG is lighter and has been sharpened.

**Reply by xiaohenian** [reply]
Thank you for your answer.I found myself in a situation like this:
https://legacy.imagemagick.org/discourse-server/viewtopic.php?f=1&t=35629&hilit=nef
But when I add a line to the "delegates.xml" file:
`<delegate decode="mynef" stealth="True" command="dcraw.exe -e -6 -w -O &quot;%o&quot; &quot;%i&quot;"/> `
Then input：
`magick mynef:FIS_0233.NEF out.jpg`
But the result was a failure：
`magick: UnableToOpenBlob 'mynef:FIS_0233.NEF': No such file or directory @ error/blob.c/OpenBlob/3537.`
I suspect my "delegates.xml" doesn't work.

**Reply by fmw42** [reply]
In what directory did you edit the delegates.xml file?  Was it the download folder or something like $MAGICK_HOME/etc/ImageMagick-7 (see https://imagemagick.org/script/resources.php).  Also did you happen to put it within a comment?

**Reply by xiaohenian** [reply]
> In what directory did you edit the delegates.xml file? Was it the download folder or something like $MAGICK_HOME/etc/ImageMagick-7 (see https://imagemagick.org/script/resources.php). Also did you happen to put it within a comment?

Thank you for your answer.The problem I found was that there is an "delegates.xml" file path, which I put in the same directory as an application like "magick.exe" that I could use.But now I have two questions:
1.When I input：
`dcraw -e -O out 3.jpg FIS_0233.NEF`
The orientation of the output image has changed (rotated).

2.I'm modifying the  “delegates.xml” file：
`<delegate decode="mynef" command="dcraw.exe -e -O &quot;%o&quot; &quot;%i&quot;"/> `
`<delegate decode="dng" command="dcraw.exe -e -O &quot;%o&quot; &quot;%i&quot;"/>`
Then input：
`magick mynef:FIS_0233.NEF out1.jpg`
`magick dng:FIS_0233.NEF out2.jpg`
But the color of the output image is different.
I want to know when I specify, or not specify, "dng:". What is the image output based on?

**Comment by snibgo** [comment]
When using the dcraw "-e" option, you are merely extracting the embedded JPEG image, so the options "-6 -w" are not relevant, and do nothing.

Adding that line to my delegates.xml works for me. I don't know why it doesn't work for you.

I suggest you remove ...
```
stealth="True"
```
... then try "magick -list delegate". That should include this line:
```
      mynef =>          "dcraw.exe -e -6 -w -O "%o" "%i"
```

**Reply by xiaohenian** [reply]
Thank you for your answer.The problem I found was that there is an "delegates.xml" file path, which I put in the same directory as an application like "magick.exe" that I could use.But now I have two questions:
1.When I input：
`dcraw -e -O out 3.jpg FIS_0233.NEF`
The orientation of the output image has changed (rotated).

2.I'm modifying the  “delegates.xml” file：
`<delegate decode="mynef" command="dcraw.exe -e -O &quot;%o&quot; &quot;%i&quot;"/> `
`<delegate decode="dng" command="dcraw.exe -e -O &quot;%o&quot; &quot;%i&quot;"/>`
Then input：
`magick mynef:FIS_0233.NEF out1.jpg`
`magick dng:FIS_0233.NEF out2.jpg`
But the color of the output image is different.
I want to know when I specify, or not specify, "dng:". What is the image output based on?

**Comment by snibgo** [comment]
dcraw will automatically apply the orientation flag, and this may rotate the image. To prevent this, include "-t 0".

> magick mynef:FIS_0233.NEF out1.jpg
magick dng:FIS_0233.NEF out2.jpg
But the color of the output image is different.

Yes. Your first command extracts the embedded JPEG image that was made by the camera. As I say above, the camera has "prettified" the image, changing colours and tones and sharpening. The second command reads the raw data which has not been prettified.

**Reply by xiaohenian** [reply]
Thank you！But in the "delegates. xml" file, my "command" is the same as "mynef" :
`dcraw.exe -e -O &quot;%o&quot; &quot;%i&quot;`

**Comment by snibgo** [comment]
And ... what? I don't understand what problem you want to solve.

**Comment by snibgo** [comment]
Your "magic -version" includes:
```
Delegates (built-in): ... raw ...
```
So processing of "DNG:" or ".NEF" files is done by that in-built raw delegate, not the external delegate given in delegates.xml, and the delegates.xml entry for "DNG:" is ignored.

**Reply by xiaohenian** [reply]
Thank you for your answer. I get it.I have one more question.Can external delegates be implemented in the IM API?Just as（C++）：
`Image img;`
`img.read("mynef:FIS_0233.NEF");`
`img.magick("jpg");`
`img.write("out.jpg");`
